### DISCUSSION TITLE
Defining bits 13 and 14 of cargo class as "potable / food / clean" and "non-potable / non-food / dirty", respectively.

### DISCUSSION THREAD

**Discussion by PikkaBird** [OpenTTD/OpenTTD#12975] [Features]
The current cargo class definitions do not offer a way to differentiate between "clean" and "dirty" cargos. This is particularly an issue for hopper and tanker cars, where it can be hard to differentiate a vehicle for carrying grain from one carrying coal, or one carrying milk from one carrying oil, without resorting to explicit cargo refits and exclusions. It also causes the perennial problem of having to explicitly handle Livestock cargos.

To that end, the proposal is to define two of the three unused Cargo Class bits as "potable / food / clean" and "non-potable / non-food / dirty". Two bits are required, rather than one, because the extended cargo bits should only ever be excluded, not required, for maximum compatibility; and some vehicles may wish to exclude one, the other, or neither (excluding both is probably a bad idea).

Default cargo classes should change as follows:

COAL 4010 Bulk, Dirty
OIL_ 4040 Liquid, Dirty
LVST 4020 Piece, Dirty
GRAI 2010 Bulk, Clean
IORE 4010 Bulk, Dirty
WHEA 2010 Bulk, Clean
FOOD 2084 Piece, Refrigerated, Clean
RUBR 4040 Liquid, Dirty
FRUT 2090 Bulk, Refrigerated, Clean
MAIZ 2010 Bulk, Clean
CORE 4010 Bulk, Dirty
WATR 2040 Liquid, Clean
SUGR 2010 Bulk, Clean

Thoughts / objections to this proposal?

**Comment by PeterN** [comment]
Don't forget all the super important toyland cargo types.

**Reply by PikkaBird** [reply]
> Don't forget all the super important toyland cargo types.

Taking this seriously for a second, I think the toyland cargos are debatable, as they're unrealistic - if "cola" and "toffee" are oil and ore analogues, but in reality edible, does that make them dirty or clean?

In any case, since these are exclusionary classes, *not* adding them to certain cargos will make those cargos continue to act as before - indeed, the new classes will only have a gameplay effect if both cargo and vehicle use them.

**Reply by PeterN** [reply]
I'm being serious. As a proposal it would be useful to know that other cargo types should or shouldn't be affected and the reasons why.

**Comment by PeterN** [comment]
Do we have any idea of where any NewGRFs use either of these bits already, and what effect changing the defaults would have on them?

**Reply by PikkaBird** [reply]
I'm not aware of any, and the spec spec-ifically says that they are reserved.

There's perhaps a case for not changing the default cargos or vehicles at all, and just defining the bits for NewGRF use - in which case, there's no changes to the game required at all, just an agreement to update the NewGRF spec.

**Reply by PeterN** [reply]
What is the impact of the default cargoes _not_ being changed? Vehicle authors will need to use the class exclude and also exclude by list of label as well?

**Reply by PikkaBird** [reply]
Since they are exclusionary classes, they are invisible unless they are used by both the cargo and the vehicle. So yes, NewGRF vehicle authors would still have to exclude the default cargos by label if they wanted them excluded (or redefine the default cargos in NewGRF to add the classes).

But really, handling the small number of default cargos is not a problem. The main benefit of the new classes would be simplifying the interaction of NewGRF vehicle sets, which may define dozens (hundreds?) of new and "interesting" wagons, with NewGRF Industry sets, which may define dozens of new and "interesting" cargos.

I suppose we also have to consider the default vehicles, and whether they should have added classes. If every climate has a grain hopper and coal/ore hoppers, it may be safe to add clean and dirty to them, respectively. Tropic has a water tanker which can be marked clean, but since the oil tanker is the only tanker in temperate and arctic, it cannot be marked dirty. Toyland, again, is hard to say.

**Reply by PeterN** [reply]
I'm in favour of not adding cargo classes to the default vehicles. There are none set at the moment as far as I know.

**Reply by 2TallTyler** [reply]
~~Default vehicles have cargo classes as of #9148. I am in favor of updating them with these new classes for consistency.~~

**Reply by PeterN** [reply]
> Default vehicles have cargo classes as of #9148.

Not actually correct, the default classes are chosen based on the default cargo type of the vehicle, they are not directly set on the vehicle.

It's also not desirable. The vanilla vehicles only have an Oil Tanker which would be considered dirty, so there'd be no clean tanker.

**Reply by 2TallTyler** [reply]
I stand corrected!

**Comment by 2TallTyler** [comment]
It's documented [in the specs](https://newgrf-specs.tt-wiki.net/wiki/Action0/Cargos#CargoClasses_.2816.29), but since it's come up in discussions I'll post it here for visibility: the special cargo used for refitting and other BAD FEATURES is bit 15, so this proposal would have no effect on NewGRFs which properly use said special cargo.

**Comment by PeterN** [comment]
I have a branch for this with changes to the default cargos ready, btw ;-)

**Comment by andythenorth** [comment]
What will we call the constants for these in NML?

**Comment by andythenorth** [comment]
Possible application of these classes to FIRS 5 cargos?

| Cargo Name               | Cargo Classes                                                |
|--------------------------|--------------------------------------------------------------|
| Acid                     | Liquid, Hazardous, **CC_DIRTY**                              |
| Aggregates               | Bulk (uncountable)                                           |
| Alcohol                  | Liquid, Express, Piece Goods (countable), **CC_CLEAN**       |
| Aluminium                | Piece Goods (countable)                                      |
| Ammonia                  | Liquid, Hazardous, **CC_DIRTY**                              |
| Beans                    | Bulk (uncountable), **CC_CLEAN**                             |
| Billets & Blooms         | Piece Goods (countable)                                      |
| Building Materials       | Piece Goods (countable), Bulk (uncountable)                  |
| Carbon Black             | Piece Goods (countable), Bulk (uncountable), Covered (weather protected), Powderized, **CC_DIRTY** |
| Cassava                  | Bulk (uncountable), **CC_CLEAN**                             |
| Cast Iron                | Piece Goods (countable)                                      |
| Cement                   | Bulk (uncountable), Covered (weather protected), Powderized, **CC_DIRTY** |
| Chemicals                | Liquid, Piece Goods (countable), Hazardous, **CC_DIRTY**     |
| China Clay (Kaolin)      | Bulk (uncountable), Liquid, Covered (weather protected)      |
| Chlorine                 | Liquid, Hazardous, **CC_DIRTY**                              |
| Clay                     | Bulk (uncountable), Covered (weather protected)              |
| Cleaning Agents          | Liquid, Piece Goods (countable)                              |
| Coal                     | Bulk (uncountable), **CC_DIRTY**                             |
| Coal Tar                 | Liquid, Hazardous, **CC_DIRTY**                              |
| Coffee                   | Piece Goods (countable), Express, **CC_CLEAN**               |
| Coke                     | Bulk (uncountable), **CC_DIRTY**                             |
| Concrete Products        | Piece Goods (countable)                                      |
| Copper                   | Piece Goods (countable)                                      |
| Copper Ore               | Bulk (uncountable), **CC_DIRTY**                             |
| Diamonds                 | Armoured                                                     |
| Edible Oil               | Liquid, Piece Goods (countable), **CC_CLEAN**                |
| Electrical Parts         | Piece Goods (countable)                                      |
| Engineering Supplies     | Express, Piece Goods (countable)                             |
| Explosives               | Express, Piece Goods (countable), Hazardous, **CC_DIRTY**    |
| Farm Supplies            | Express, Piece Goods (countable)                             |
| Ferroalloys              | Bulk (uncountable), Piece Goods (countable)                  |
| Fertiliser               | Piece Goods (countable), Bulk (uncountable), Covered (weather protected) |
| Fish                     | Express, Refrigerated, **CC_CLEAN**                          |
| Food                     | Refrigerated, Express, **CC_CLEAN**                          |
| Forgings & Castings      | Piece Goods (countable)                                      |
| Fruit                    | Express, Piece Goods (countable), Refrigerated, **CC_CLEAN** |
| Glass                    | Piece Goods (countable)                                      |
| Goods                    | Express, Piece Goods (countable)                             |
| Grain                    | Bulk (uncountable), **CC_CLEAN**                             |
| Hardware                 | Piece Goods (countable), Express                             |
| Iron Ore                 | Bulk (uncountable)                                           |
| Limestone                | Bulk (uncountable)                                           |
| Livestock                | Piece Goods (countable), **CC_DIRTY**                        |
| Logs                     | Piece Goods (countable)                                      |
| Mail                     | Mail                                                         |
| Maize                    | Bulk (uncountable), **CC_CLEAN**                             |
| Manganese                | Bulk (uncountable), **CC_DIRTY**                             |
| Merchant Bar             | Piece Goods (countable)                                      |
| Milk                     | Express, Liquid, Refrigerated, **CC_CLEAN**                  |
| Nitrates                 | Bulk (uncountable), **CC_DIRTY**                             |
| Nitrogen                 | Liquid                                                       |
| Nuts                     | Piece Goods (countable), **CC_CLEAN**                        |
| Oil                      | Liquid, **CC_DIRTY**                                         |
| Oxygen                   | Liquid                                                       |
| Paints & Coatings        | Liquid, Piece Goods (countable), **CC_DIRTY**                |
| Paper                    | Piece Goods (countable), **CC_CLEAN**                        |
| Peat                     | Bulk (uncountable)                                           |
| Petroleum Fuels          | Liquid, **CC_DIRTY**                                         |
| Phosphate                | Bulk (uncountable)                                           |
| Pig Iron                 | Piece Goods (countable)                                      |
| Pipework                 | Piece Goods (countable), Oversized                           |
| Potash                   | Bulk (uncountable)                                           |
| Pumps & Valves           | Piece Goods (countable)                                      |
| Pyrite Ore               | Bulk (uncountable), **CC_DIRTY**                             |
| Quicklime                | Bulk (uncountable), Covered (weather protected), **CC_DIRTY**|
| Rebar                    | Piece Goods (countable)                                      |
| Rubber                   | Liquid                                                       |
| Salt                     | Bulk (uncountable), Covered (weather protected)              |
| Sand                     | Bulk (uncountable), Covered (weather protected)              |
| Scrap Metal              | Bulk (uncountable), Not Pourable, **CC_DIRTY**               |
| Seals, Hoses & Belts     | Piece Goods (countable)                                      |
| Slag                     | Bulk (uncountable), **CC_DIRTY**                             |
| Soda Ash                 | Bulk (uncountable), Covered (weather protected)              |
| Sodium Hydroxide (Lye)   | Liquid, Hazardous, **CC_DIRTY**                              |
| Steel                    | Piece Goods (countable)                                      |
| Steel Ingots             | Piece Goods (countable)                                      |
| Steel Pipe               | Piece Goods (countable), Oversized                           |
| Steel Sheet & Strip      | Piece Goods (countable)                                      |
| Steel Slab               | Piece Goods (countable)                                      |
| Steel Tube               | Piece Goods (countable)                                      |
| Steel Wire Rod           | Piece Goods (countable)                                      |
| Structural Steel         | Piece Goods (countable)                                      |
| Sulphur                  | Bulk (uncountable), Liquid, Covered (weather protected)      |
| Timber                   | Bulk (uncountable), Piece Goods (countable)                  |
| Tyre Cord                | Piece Goods (countable)                                      |
| Tyres                    | Piece Goods (countable)                                      |
| Vehicle Bodies           | Piece Goods (countable), Oversized                           |
| Vehicle Engines          | Piece Goods (countable)                                      |
| Vehicle Parts            | Piece Goods (countable)                                      |
| Vehicles                 | Piece Goods (countable), Oversized                           |
| Welding Consumables      | Liquid, Express, Piece Goods (countable)                     |
| Wool                     | Piece Goods (countable), Covered (weather protected), **CC_CLEAN** |
| Zinc                     | Piece Goods (countable)                                      |

**Comment by andythenorth** [comment]
Having tried to rationalise applying these to FIRS, I'm not sure what I'd actually do.  

Currently as industry set author it looks like it's wide open to nitpicks from consumers and vehicle grf authors.

I only have confusion to offer, sorry 🫤 

As vehicle author:

- CC_CLEAN seems like it's CC_FOOD or CC_POTABLE or CC_EDIBLE, and should be an include not exclude. I know why we favour exclude not include, and was part of the original working out of how to handle extended classes with exclude only, but 'food vehicle' is not sufficiently handled by 'exclude dirty' - which means using labels, but that might be fine

- CC_DIRTY, I'm not sure how that would be used generally.  In FIRS, I can't see how to avoid marking all CC_BULK and CC_LIQUID as also either CC_DIRTY or CC_CLEAN currently.

That said, it's relatively easy to apply this to tankers, as IRL tankers tend to be 
- food grade (exclude CC_DIRTY)
- petrochemical (exclude CC_CLEAN)

But I can't see how to usefully apply it to hoppers, opens, covered hoppers or vans.  

My confusion isn't about the bits, it's about what the intent is, and I'm probably confused by being both FIRS author and vehicle set author. 

--- EDIT ---

To try and resolve it to a single question.

- ignoring vehicles, because vehicles can and should always use labels for explicit control
- **for cargos, what's my heuristic for applying CC_DIRTY to a specific cargo (or not)?**

**Reply by PikkaBird** [reply]
For IH, I should think sometimes using includes for secondary labels on wagons would not be a problem. You do, after all, have a thousand wagons to play with. As long as some generic wagons exist for carrying unanticipated cargos.

My feeling for cargo labelling is this, and bearing in mind other GRF authors may interpret differently, and that's fine:

CC_CLEAN should be set on cargos which should not be exposed to contaminants. In Sunshine, this is Grain, Fish, Fruit, and Food.
CC_DIRTY should be set on cargos which contaminate. In Sunshine, this is Oil, Fuel Oil, Coal ,Iron Ore, Stone.

Cargos which are neither particularly dirty, nor need protection from dirty cargos, get neither label. In Sunshine; Fiber Crops, Steel, Lumber, Wood, and the generic "Chemicals" and "Supplies" cargos.

For vehicles, I exclude CC_CLEAN from hoppers and tankers designed for dirty industrial cargos, and CC_DIRTY from hoppers and tankers designed for food and potable liquids.

**Reply by andythenorth** [reply]
Would it make sense for Fish to be both CC_CLEAN and CC_DIRTY?

**Reply by PikkaBird** [reply]
> Would it make sense for Fish to be both CC_CLEAN and CC_DIRTY?

I probably wouldn't combine the two bits, as it seems plausible vehicle sets might not include any vehicles that don't exclude one or the other (eg, only have two tankers, one that excludes clean, and one that excludes dirty).

If the goal for your vehicle set is "fish can only be carried in fish vehicles, which can carry only fish," then that's a job for explicit cargos.

**Reply by andythenorth** [reply]
Yup, so just sticking to the cargo set author side, and keeping vehicle behaviour outwith that.

**IRL, rubber and plastics both:**
- contaminate
- must be protected from contamination

Which would prevail?

**Do any piece goods get CC_DIRTY?**

**Reply by PikkaBird** [reply]
> Which would prevail?

Personally, I would tag [raw] rubber as dirty, and plastics as neither. If your intention is that they must go only in wagons designated for _this specific cargo_, that is outside the scope and the spirit of the cargo class system. Like I said, you can do that with explicit cargo fits in your vehicle set.
 
> **Do any piece goods get CC_DIRTY?**

By convention, I'd suggest Piece + CC_DIRTY imply a Livestock cargo, or at least something that is carried in livestock-esque (ventilated etc) wagons.

**Reply by andythenorth** [reply]
Something isn't clicking for me, but I do think this is a good idea. 😁 

From the _industry author_ side, it's best to not try and rationalise what vehicles might do with a cargo.  I just try and reflect (approximately) IRL properties.

For the _vehicle author_, I'm starting to think 'include' is better for CC_CLEAN than 'exclude'.  

Both CC_DIRTY and CC_CLEAN have the same issue with cargos predating these classes.

The problem I've most consistently had with vehicles is detecting the food-type cargos, whether piece, bulk or liquid.

Currently I do that by label, for example:
-  includes https://github.com/andythenorth/polar-fox/blob/main/src/constants.py#L236
- excludes https://github.com/andythenorth/polar-fox/blob/main/src/constants.py#L294

So CC_CLEAN makes sense, but my inclination is to use it on an include list.  Either as OR, or AND with CC_BULK, CC_LIQUID or CC_PIECE_GOODS

Snippet of my typical current code for a food tanker / van / fridge van / covered hopper.
```            
refittable_cargo_classes:       bitmask();
non_refittable_cargo_classes:   bitmask(); // don't set non-refittable classes, increases likelihood of breaking cargo support
cargo_allow_refit:              [BEER,EOIL,FOOD,MILK,WATR];
```

Sidetrack: the NML docs don't have any of the advice from the nfo docs about include / exclude combinations.  Can't even remember the details, but it was definitely a thing at the time 😨 

- https://newgrf-specs.tt-wiki.net/wiki/NML:Cargos#Cargo_classes
- https://newgrf-specs.tt-wiki.net/wiki/Action0/Cargos#CargoClasses_.2816.29

**Reply by PikkaBird** [reply]
I did notice that lack of info in the NML docs. I think the advice sums up as:

Every cargo should have at least one of the basic (pax, mail, express, armored, bulk, piece, liquid) classes set.
Every vehicle should have at least one basic class on its include list.
The additional classes are optional, for fine-tuning vehicle selection.
Vehicles should sparingly exclude these additional classes, only if a non-exclusionary vehicle is available (eg, a flat wagon should only exclude CC_OVERSIZED if another vehicle is available that does not).

Creating a clean-only hopper would entail setting the include to CC_CLEAN, and the exclude to [all the basic classes except CC_BULK]. And it's an option I can see working for IH, since you have enough generic hopper types that other cargos would still be transportable. I guess it calls for experimentation? Just setting the bits for cargos and vehicles doesn't require any changes to OpenTTD...

**Reply by PeterN** [reply]
> ```
> refittable_cargo_classes:       bitmask();
> non_refittable_cargo_classes:   bitmask(); // don't set non-refittable classes, increases likelihood of breaking cargo support
> cargo_allow_refit:              [BEER,EOIL,FOOD,MILK,WATR];
> ```

Breaking cargo support was a thing if you used the old original refit mask property, which would XOR with the `include classes` and `exclude classes` result. Because it was XOR it could lead to opposites happening.

Because labels are 'well-defined', the later `include label` and `exclude label` properties are applied after the classes are done, and will simplify add or remove as desired. This won't lead to unexpected broken cargo types.

**Reply by andythenorth** [reply]
> I guess it calls for experimentation? Just setting the bits for cargos and vehicles doesn't require any changes to OpenTTD...

I could patch NML for this locally.  Not right now, but yes.

**Comment by andythenorth** [comment]
> 09:26]peter1138: Cargo class labels?
> [09:26]peter1138: (Badges)
> [09:26]Pruple: badges solve everything
> [09:33]peter1138: Well...
> [09:33]peter1138: Not quite.

> [09:41]andythenorth: no but cargo class badges 😛
> [09:41]andythenorth: I didn't want to derail the discussion
> [09:42]andythenorth: but I've got a bunch of vehicles for 'food and farm'
> [09:42]andythenorth: and another set for 'metals'
> [09:42]andythenorth: and another set for 'cargos that are long'
> [09:42]andythenorth: combinatorial classes aren't quite capturing them 😛
> [09:43]andythenorth: I get a cheat code because I wrote my own shared python lib to deal with this at one level of abstraction
> [09:43]andythenorth: but it's just many many lists of labels 😛

If I was starting from scratch, as cargo author I'd:

- use the basic classes, which relate to physical properties: liquid, bulk, piece
- add a set of 'like-for-like' attributes (badges?) which relate to some sort of standard industrial classifications: foodstuffs, petrochemicals, building materials, ores, manufactured goods, metal stock
- add a set of 'handling' attributes (badges?): needs weather protection, can be poured under gravity, will flow under air pressure, etc

These would inevitably fragment and lead to multiple standardisation specs, but that's usual and fine.

**Reply by PeterN** [reply]
Ignore badges, they are designed for users for cosmetics primarily. There is a varact2 test to allow NewGRFs to do things based on them, but the game itself will not use them to control anything.

A completely dynamic cargo-class-label system may be feasible to add, with an include/exclude cargo-class-label list. This is basically cargo classes as now, but not limited to the fixed 16 bits. In which case this proposal would just be defining a convention rather than using up precious bits...

Vehicle variable 0x42 and 0x47, and callback CBID_VEHICLE_REFIT_COST use the 16 bit cargo classes though, so not clear how to work with that. And some of the core code tests for CC_PASSENGERS/CC_MAIL/CC_EXPRESS/CC_ARMOURED.

**Reply by andythenorth** [reply]
This could work...

**Reply by andythenorth** [reply]
My inclination is that a total free-for-all won't go well, and that the current core classes are generally valid.

- Bulk
- Piece
- Liquids
- Passengers
- Mail

The following are weird, and useless for most vehicle set cases, but are established fact in core game
- Express
- Armoured

The other bits have always been weird.  Some are useful, some are not

- Hazardous
- Covered
- Oversized
- Powderized
- Not Pourable
- Special

Not sure what to make of that.  

- keep the current classes as established fact, and don't provoke pitchforks by touching them, but add `cargo-class-label` as separate system?
- keep a subset of the current classes, and augment with `cargo-class-label` system?
- reimplement all classes as `cargo-class-label` system?

**Reply by andythenorth** [reply]
Tried a left-turn
- some of the current difficulties with classes arise for trying to use classes for cargos which are
  - like food (fish, milk, alcohol)
  - like grain (wheat, maize, beans, oilseeds)
  - like cement (carbon black, quicklime)
  - like salt (potash, soda ash)
  - like acid (ammonia, cleaning agents, sodium hydroxide)

As a vehicle set author, it's currently easier/more predictable to just handle explicit labels for some cases, such as

- covered hoppers (grain, salt)
- silo tankers with pressure discharge (cement etc)
- food tankers (milk, alcohol)
- food wagons (fish, food, alcohol)

The core game has quite a nice range of cargos and vehicle types.  It didn't include anything for covered hoppers or silos, but the basic range is pretty good.  

Although it diverges from this proposal, I was curious if anything good might come of letting industry grf authors have a new action 0 prop `handled like`, which takes _one_ existing cargo label.  

So e.g. the cargo action 0 for 'beans' defines label `BEAN` and `handled_like: GRAI` and then any vehicle refitting `GRAI` also refits `BEAN`.  

Nothing else changes in vars, action 3 etc - I envisage this being possibly as simple as extending the list in train prop 0x2C with any `handled-like` labels that are found in the action 0 prop 0x2C definition.

**I didn't think through** how that might work with indexes into CTTs, and/or what happens if `GRAI` is not defined in the current game.  That might be an impediment to this idea.

I tested applying that to FIRS, using just labels from the core game.  
- notably there's nothing for 'powders' or 'minerals that need weather protection'
- there's nothing for liquid chemicals, but they could probably be handled same as oil?
- almost everything else could be covered from core game cargos

To a large extent, classes could also cover this (as the use of "paper" for nearly all piece goods shows), but classes are basically playing paper-scissors-stone between cargo authors and vehicle set authors.

"Acts like grain" is just less to think about, and mirrors that, broadly, any vehicle author who wants accurate control needs to use labels anyway, however this would significantly reduce the burden for vehicle set authors when large sets like FIRS release lots of new labels.

```
| Name                  | Maps to    |
|-----------------------|------------|
| Acid                  |            |
| Aggregates            | Iron Ore   |
| Alcohol               | Cola       |
| Aluminium             | Steel      |
| Ammonia               |            |
| Beans                 | Grain      |
| Billets & Blooms      | Steel      |
| Building Materials    | Paper      |
| Carbon Black          |            |
| Cassava               | Fruit      |
| Cast Iron             | Steel      |
| Cement                |            |
| Chemicals             |            |
| China Clay            |            |
| Chlorine              |            |
| Clay                  |            |
| Cleaning Agents       |            |
| Coal                  | Coal       |
| Coal Tar              | Oil        |
| Coffee                | Fruit      |
| Coke                  | Coal       |
| Concrete Products     | Paper      |
| Copper                | Steel      |
| Copper Ore            | Copper Ore |
| Diamonds              | Diamonds   |
| Edible Oil            | Water      |
| Electrical Parts      | Paper      |
| Engineering Supplies  | Goods      |
| Explosives            | Valuables  |
| Farm Supplies         | Goods      |
| Ferroalloys           | Iron Ore   |
| Fertiliser            |            |
| Fish                  | Food       |
| Food                  | Food       |
| Forgings & Castings   | Steel      |
| Fruit                 | Fruit      |
| Glass                 | Paper      |
| Goods                 | Goods      |
| Grain                 | Grain      |
| Hardware              | Paper      |
| Iron Ore              | Iron Ore   |
| Limestone             | Iron Ore   |
| Livestock             | Livestock  |
| Logs                  | Wood       |
| Mail                  | Mail       |
| Maize                 | Maize      |
| Manganese             | Iron Ore   |
| Merchant Bar          | Steel      |
| Milk                  | Water      |
| Nitrates              |            |
| Nitrogen              |            |
| Nuts                  | Fruit      |
| Oil                   | Oil        |
| Oxygen                |            |
| Paints & Coatings     |            |
| Paper                 | Paper      |
| Passengers            | Passengers |
| Peat                  |            |
| Petroleum Fuels       | Oil        |
| Phosphate             | Iron Ore   |
| Pig Iron              | Steel      |
| Pipework              | Paper      |
| Potash                |            |
| Pumps & Valves        | Paper      |
| Pyrite Ore            | Iron Ore   |
| Quicklime             |            |
| Rebar                 | Steel      |
| Rubber                | Rubber     |
| Salt                  |            |
| Sand                  | Iron Ore   |
| Scrap Metal           | Iron Ore   |
| Seals, Hoses & Belts  | Paper      |
| Slag                  | Iron Ore   |
| Soda Ash              |            |
| Sodium Hydroxide      |            |
| Steel                 | Steel      |
| Steel Ingots          | Steel      |
| Steel Pipe            | Steel      |
| Steel Sheet & Strip   | Steel      |
| Steel Slab            | Steel      |
| Steel Tube            | Steel      |
| Steel Wire Rod        | Steel      |
| Structural Steel      | Steel      |
| Sulphur               |            |
| Timber                |            |
| Tyre Cord             | Steel      |
| Tyres                 | Paper      |
| Vehicle Bodies        |            |
| Vehicle Engines       | Paper      |
| Vehicle Parts         | Paper      |
| Vehicles              |            |
| Welding Consumables   | Goods      |
| Wool                  | Fruit      |
| Zinc                  | Steel      |
```

**Comment by andythenorth** [comment]
The most compelling of the 6 proposals I have so far 😛 

The new bits, per this proposal and some redefinitions / docs changes.

This isn't arbitrary whim, I've worked out the implications by checking all the cargos listed at https://newgrf-specs.tt-wiki.net/wiki/CargoTypes#Cargo_Labels 

This is the least-surface-area change I've been able to come up with, apart from 'do nothing' 😁   I don't love the exclude route, I think almost zero vehicle grf authors understand it or use it, but if that's the preference, then I think the changes below would work and would be positive impact for low effort (and low pitchfork count).

```
0    1       Passengers                          passenger vehicle                  Keep - hard-coded dependencies in OpenTTD - no pitchforks
1    2       Mail                                closed or mail wagon               Keep - hard-coded dependencies in OpenTTD - no pitchforks
2    4       Express                             closed or mail wagon               Keep - hard-coded dependencies in OpenTTD - no pitchforks
3    8       Armored                             armored or mail wagon              Keep - hard-coded dependencies in OpenTTD - no pitchforks
4    10      Bulk (Uncountable)                  open or hopper wagon               Keep - no pitchforks 
5    20      Piece Goods (Countable)             closed or open wagon               Keep - but it's really 'general cargo' and many cargos should be setting this - some pitchforks
6    40      Liquid                              tank wagon                         Keep - no pitchforks
7    80      Refrigerated                        refrigerated wagon                 Slightly silly, I think this is better as packaged food (controlled temperature, humidity etc) - but might not be worth the candle - some pitchforks
8    100     Hazardous                           unknown                            Silly class, used for gas and radioactive cargos with completely different wagons.  Redefine as CC_GAS - limited pitchforks
9    200     Covered (weather protected)         closed wagon                       Keep - but change advice - this really only makes sense as modifier on bulk - some pitchforks
10   400     Oversized                           stake/flatbed wagon                Keep - but change advice - this really only makes sense as a modifier on piece goods for 'flatbed freight' or 'large pieces' - limited pitchforks
11   800     Powderized (moist protected)        powder/silo wagon                  Keep - drop 'moisture protected', irrelevant; restore 'flows under pressure'.  Make specific class, not an exclude, recommend fallback to piece not bulk - limited pitchforks
12   1000    Not Pourable                        open wagon, but not hopper wagon   Silly class, hardly used, convert to 'requires open top' for cargos that should *not* generally be covered - limited pitchforks
13   2000    reserved.                                                              CC_FOOD_GRADE / CC_CLEAN / CC_POTABLE - no pitchforks
14   4000    reserved.                                                              CC_NON_FOOD_GRADE / CC_DIRTY / CC_NON_POTABLE - no pitchforks
15   8000    special                                                                Keep - daft, but hard-coded dependencies in OpenTTD - no pitchforks

```

**Reply by andythenorth** [reply]
My preferred option right now.

- redefine 4 or 5 classes (wiki advice, class names, or both)
- do bits 13 and 14 for potable / non-potable
- deprecate oversized and not-pourable as daft, mostly unused, mostly useless 😉 

```
0    1       Passengers                          passenger vehicle                  Keep - hard-coded dependencies in OpenTTD - no pitchforks
1    2       Mail                                closed or mail wagon               Keep - hard-coded dependencies in OpenTTD - no pitchforks
2    4       Express                             closed or mail wagon               Keep - hard-coded dependencies in OpenTTD - no pitchforks
3    8       Armored                             armored or mail wagon              Keep - hard-coded dependencies in OpenTTD - no pitchforks
4    10      Bulk (Uncountable)                  open or hopper wagon               Keep - adjust advice - this is more useful as 'bulk open' - some pitchforks
5    20      Piece Goods (Countable)             closed or open wagon               Keep - but it's really 'general cargo' and many cargos should be setting this - some pitchforks
6    40      Liquid                              tank wagon                         Keep - no pitchforks
7    80      Refrigerated                        refrigerated wagon                 Slightly silly, I think this is better as packaged food (controlled temperature, humidity etc) - but might not be worth the candle - some pitchforks
8    100     Hazardous                           unknown                            Silly class, used for gas and radioactive cargos with completely different wagons.  Redefine as CC_GAS - limited pitchforks
9    200     Covered (weather protected)         closed wagon                       Keep - adjust advice - this is more useful as 'bulk covered' - limited pitchforks
10   400     Oversized                           stake/flatbed wagon                Deprecate, hardly used, makes no sense - low pitchforks
11   800     Powderized (moist protected)        powder/silo wagon                  Keep - drop 'moisture protected', irrelevant; restore 'flows under pressure'.  Make specific class, not an exclude, recommend fallback to piece not bulk - limited pitchforks
12   1000    Not Pourable                        open wagon, but not hopper wagon   Deprecate, hardly used, makes no sense - limited pitchforks
13   2000    reserved.                                                              CC_FOOD_GRADE / CC_CLEAN / CC_POTABLE - no pitchforks
14   4000    reserved.                                                              CC_NON_FOOD_GRADE / CC_DIRTY / CC_NON_POTABLE - no pitchforks
15   8000    special                                                                Keep - daft, but hard-coded dependencies in OpenTTD - no pitchforks
```

**Comment by andythenorth** [comment]
I've done a bit of reviewing existing cargos at https://newgrf-specs.tt-wiki.net/wiki/CargoTypes#Cargo_Labels

I think proposed bits 13 and 14 will work well.

**New bits**
- adding bits 13 and 14 makes sense to me
- to me, they make most sense as `food-grade` and `non-food-grade`, but YMMV
- (DRAFT) there are some docs here that show with more examples how bits 13 and 14 could work (docs also contain other unrelated proposals) [https://grf.farm/polar_fox/cargo_classes_A.html#CC_FOOD_GRADE](https://grf.farm/polar_fox/cargo_classes_A.html#CC_FOOD_GRADE)

**Unrelated comments**
- it would be helpful to have bit 9 `covered` set on GRAI, WHEA, MAIZ, SUGR
- FRUT being `bulk` but not `piece goods` is odd
- GOOD being `express` but not `piece goods` is odd 
- `refrigerated` looks relatively useless for refits, most cargos also set `piece` and could now set bit 13; Pikka says it remains useful for switching sprites though.

But changing the base game cargos is probably unwise eh?

**For future travellers**

- I do have a sense of previously burnt fingers here.  Exclusion classes didn't work last time either, with powderised / covered / oversized / non-pourable, which are broadly not used or useful.  Maybe this time will be different? 😄

**Comment by andythenorth** [comment]
In terms of objections/comments, the forum post has been up for more than 2 weeks with no comments. https://www.tt-forums.net/viewtopic.php?t=91942

I've asked multiple times in Discord #add-on-development, which tends to be the most active place for newgrf discussion these days; nobody has objected to bits 13 and 14.  Discussion so far suggests that active newgrf authors:
- don't really pay much attention to cargo classes
- possibly just cargo cult them from existing examples

So I think there's no obvious problem going ahead with these bits.

**Comment by andythenorth** [comment]
OpenTTD PR https://github.com/OpenTTD/OpenTTD/pull/12979
NML PR https://github.com/OpenTTD/nml/pull/343

**Comment by michicc** [comment]
Related, but different: PR #13090

**Comment by andythenorth** [comment]
OpenTTD and NML PRs merged, grf docs updated.  

Think we can close this discussion? 😄
